[WIP] July 2018 Update

Dabomstew opened the pull request:
**Second currency**
- [x] Database fields
- [x] Second currency amount displayed when you use !points (and maybe an alias of its name as well)
- [x] Second currency expires at the start of the month 2 after the one it was earned (so if earned from August 1-31, would expire in early October)
- [x] Purchase packs with second currency (savings closest to expiration are used first)
- [x] Bets pay out in second currency
- [x] Sets pay out in second currency

**System to support Anniversary packs**
- [x] Database field for event tokens
- [x] Each pack has a chance to replace a card with an event token which is credited to account immediately
- [x] Event tokens shown in visual display of pack but do not contribute towards keep/DE list
- [x] Database settings for name, image, possibly other aspects

**God image request system**
- [x] !godimage command with the following:
  - [x] request/requestglobal <waifu id> - places the request from user side, does preliminary checks (url must be valid from imgur or some other trusted host, mime type, possibly aspect ratio). If requestglobal is used, submits the image as a candidate to be used for all copies, otherwise just your own. This command should send a notification to a new `adminDiscordHook` when successfully invoked. Should also warn user if image submitted is >= 2MB so they can choose to adjust it rather than being unwittingly subjected to the auto downscale.
  - [x] list - check your own pending requests
  - [x] cancel <id> - cancels a particular request
  - [x] queue - lists pending request ID numbers
  - [x] check <id> - shows the details of a particular request
  - [x] acceptsingle <id> - accepts a particular request, but only as the image for that user's copy, even if the user requested a global change.
  - [x] acceptglobal <id> - accepts a request for a global change and does the global change. If used on a non-global request, errors out.
- [x] When a request is accepted,
  - [x] download it (and error out if download fails)
  - [x] double check that the properties tested on submission haven't become invalid
  - [x] downscale to under 2mb while above 2mb
  - [x] reupload to our own image host
  - [x] send appropriate whisper notification
- [x] Requests only usable on cards with a promoted god copy present in hand
- [x] Requests cancelled when (send whisper to let the user know as well)
  - [x] copy is disenchanted
  - [x] copy is traded

**Profile pics / special cards cleanup**
- [x] Block trading specials entirely
- [x] Show promoted rarity if a promoted copy of profile favorite is present in hand
- [x] Make a flag for certain cards to not be allowed as profile favorite at all
- [x] Remove card from profile favorite if:
  - [x] Promo/special is disenchanted
  - [x] Promo is traded away
  - [x] Card is raritychanged to promo or special and not present in hand

**Centralized free packs system**
- [x] Database table for freepacks inventory
- [x] Freepacks command:
  - [x] base command (or list) lists your current inventory of free packs, like current !bet packs
  - [x] !freepacks open/redeem <packname> to cash one in, so as to remove the clutter in the token system
- [x] Any remaining bet packs (tier 6/7, probably) are sent to freepacks system
- [x] Packs given out by freebie can be sent there.
- [x] Add system to auto-dispense a certain free pack upon getting a "meme pack" situation (75/625 jumbo, 420 standard) and notify the user of such

xorhash commented:
AFAICT missing:

Moving the 420 standards and 75/625 jumbos into the freepacks system; add notification if this case happens (especially if obscured by bounties breaking the "natural" de value).

---

A few notes:

> Database field for event tokens

I know that the discussion thus far has called them "event tokens", but this collides with the "token" for !redeem. While "token" is not established parlance in the community, it'd probably still at least help code clarity to rename the currency from "event tokens" to maybe "event coins" or something.

> request/requestglobal - places the request from user side, does preliminary checks (url must be valid from imgur or some other trusted host, mime type, possibly aspect ratio).

Resolution should probably also be capped at some reasonable amount.

Moreover, I think request/requestglobal sound suboptimal as command names. I see it as rather likely that people will use `!request global` or even forget to supply "global" at all; I assume people will often try to force their request for everybody in the first place. At least make `!request global` alias with `!requestglobal`. (IMO the default behavior should be swapped and be made that the default is global unless requested otherwise because of that; but that's a policy decision and thus my input is irrelevant.)

The command name is also possibly misleading, especially to new people (who might not even know about promotion). Even if it's a bit wordy, maybe `!godimage` or `!godimagechange` might be a better fit. </bikeshedding>

> downscale to under 2mb while above 2mb

I'd suggest you warn users if downscaling is imminent. Depending on your algorithm, the result might be worse than what a user could achieve by clever cropping, so users should be given a chance to fix it themselves.

> Requests cancelled when [...] traded

You may want to send a whisper notification or a second message on the trade in this case in particular; people might have a request pending but also use the card for circle trades or self trades.

Dabomstew commented:
@xorhash 

I have addressed points 1, 3 and 4 with checklist adjustments.

As for point 2, the command for making the requests is meant to be `!imagechange request[global]`, `!request` doesn't factor into it. Changing that subcommand name to something else is possible, though I have no idea what said change would be to. `!godimage` for the overall command name also isn't a terrible idea.

wherewedroppinboys commented:
how about `!changepic [id]` and `!changepic [id] global`

Dabomstew commented:
The plan is definitely to have everything for that system under one command to keep it encapsulated. `!godimage change [id]` / `!godimage changeglobal [id]` could work

xorhash commented on main.py:
```diff
@@ -4835,7 +4862,7 @@ def do_command(self, command, args, sender, channel, tags, isWhisper=False):
                     return
 
-                if waifu['base_rarity'] >= int(config['numNormalRarities']):
+                if waifu['base_rarity'] == int(config['numNormalRarities']):
```
Do you really want to change this? I know the message says "special", but is there a legitimate use case to un-promo promos?

Dabomstew replied:
Re-adding them to packs, which functions fine by just having the people who already own the card keep their copy at promo. Just need to disable the stars for promo rarity.

xorhash replied:
That'd be a (possibly massive) hand space penalty for people who rely on promos staying promos, though. Is that really a guarantee you can just drop like this?

Dabomstew replied:
Re-read my comment - any existing promo copies would stay at promo for the duration that the card was back in packs.

xorhash commented:
Re abf607e8e as a whole: I suspect you're trying to keep things flexible, but I feel like it might be a better idea to swap Special and Promo in the db so that you can do a `>=` check with the Special rarity, rather than having one rarity below Promo be very special.

Dabomstew commented:
Ultimately that depends whether we want [one super special rarity and potentially infinitely more pseudo-special rarities] or the opposite, will wait and see what Maren has to say.

xorhash commented:
Also mentioning issue #61 here since this PR is planning to fix it.

xorhash commented on main.py:
```diff
@@ -5031,6 +5026,102 @@ def do_command(self, command, args, sender, channel, tags, isWhisper=False):
+                        return
+
+                    with db.cursor() as cur:
+                        cur.execute("SELECT COUNT(*) FROM has_waifu WHERE userid = %s AND waifuid = %s AND rarity = %s", [tags['user-id'], waifuid, godRarity])
```
I think this assumes that there's only one God card in a player's hand. Given this might not be necessarily the case (there's the potential special case of a duplicated God via Miku's Mercy), you may want to error out if the count is > 1.

Dabomstew replied:
Way too much of this code has to rely on gods being singular for it to function sanely. I think Miku's Mercy will have to not apply to Gods, regardless of Maren's generosity.

Dabomstew replied:
Not apply to Gods where it would cause a second copy to exist, that is.

Dabomstew replied:
(Basically, I'm not rewriting the whole thing just to cover one extremely rare and extremely dumb edge case)

xorhash commented on main.py:
```diff
@@ -4957,7 +4978,7 @@ def do_command(self, command, args, sender, channel, tags, isWhisper=False):
                         newFav = int(args[1])
                     except ValueError:
-                        self.message(channel, args[1] + " is not a number. Please try again.");
+                        self.message(channel, args[1] + " is not a number. Please try again.")
```
Side note: Why is this using an `int()` check? Other places, e.g. `bounty <ID> <amount>`, just pretend there's a syntax error when the waifu ID isn't found or not lookupable. You do a `getWaifuById()` shortly after anyway.

Dabomstew replied:
You're right, but refactors aren't really the goal of this update - I was just cleaning up low-hanging fruit reported to me by VS Code's linter.

xorhash commented on main.py:
```diff
@@ -5031,6 +5026,102 @@ def do_command(self, command, args, sender, channel, tags, isWhisper=False):
+                                "username": "WTCG Admin", 
+                                "content" : "{user} requested {type} image change for [{id}] {name} to <{image}>!\nUse `!godimage check {id}` in any chat to check it.".format(**discordArgs)
+                            }
+                            sendAdminDiscordAlert(discordbody)
```
Given the regular `sendDiscordAlert` function is in a separate thread, is there any reason you're *not* putting sendAdminDiscordAlert in its own thread? Unless I misunderstand how the requests module works, the bot could hang here if the Discord API is unreachable.

Dabomstew replied:
Oversight, actually - this should definitely be threaded, will fix later

xorhash commented on main.py:
```diff
@@ -5087,8 +5087,8 @@ def do_command(self, command, args, sender, channel, tags, isWhisper=False):
-                            waifuInfo = getWaifuById(waifuid)
-                            discordArgs = {"user": tags['display-name'], "id": waifuid, "name": waifuInfo["name"], "image": args[2], "type": "a global" if do_global else "an"}
+                            waifu = getWaifuById(waifuid)
+                            discordArgs = {"user": tags['display-name'], "id": waifuid, "name": waifu["name"], "image": args[2], "type": "a global" if do_global else "an"}
```
Wording: Maybe it should be "an individual"/"a global" rather than just implicitly stating "an image change" for an individual image change. The more explicit wording might prevent misunderstandings, especially if hastily reading the message on the go.

xorhash commented on main.py:
```diff
@@ -3695,10 +3695,58 @@ def do_command(self, command, args, sender, channel, tags, isWhisper=False):
+                            return
+
+                        # can they actually open it?
+                        cur.execute("SELECT COUNT(*) FROM boosters_opened WHERE userid = %s AND status = 'open'",
```
I know this is a feature update, but the [wisdom of StackOverflow](https://stackoverflow.com/a/5264695) says that an `EXISTS` subquery [could be faster up to 25%](https://stackoverflow.com/a/5264757). Given how big `boosters_opened` will be getting since nobody ever seems to delete from there (and things depend on that!), this may be worthwhile to introduce as a pattern in the entire codebase.

xorhash commented on main.py:
```diff
@@ -3695,10 +3695,58 @@ def do_command(self, command, args, sender, channel, tags, isWhisper=False):
+                            if checkHandUpgrade(tags['user-id']):
+                                messageForHandUpgrade(tags['user-id'], tags['display-name'], self, channel, isWhisper)
+                            cur.execute("UPDATE freepacks SET remaining = remaining - 1 WHERE userid = %s AND boostername = %s", [tags['user-id'], args[1]])
+                            self.message(channel, "%s, you open a free %s booster: %s/booster?user=%s" % (tags['display-name'], result[1], config["siteHost"], sender), isWhisper)
```
Wording: It may be worth reconsidering the word "free".

On one hand, it makes clear that no money has been spent. On the other hand, it may feel like they might be getting less value, similar to how people seemed to need to be reaffirmed that bet packs contribute towards pity.

xorhash commented on main.py:
```diff
@@ -4257,6 +4257,41 @@ def do_command(self, command, args, sender, channel, tags, isWhisper=False):
                 return
+            if command == "givefreepack" and sender in superadmins:
+                if len(args) < 2:
+                    self.message(channel, "Usage: !givefreepack <username> <booster name> [<amount>] (default 1)", isWhisper)
```
Wording: Maybe put the "(default 1)" in the brackets, i.e.:

    Usage: !givefreepack <username> <booster name> [<amount> (default 1)]

This seems clearer to me.

xorhash commented on main.py:
```diff
@@ -4257,6 +4257,41 @@ def do_command(self, command, args, sender, channel, tags, isWhisper=False):
+
+                    giveFreeBooster(userData[0], args[1], amount)
+                    if amount > 1:
+                        self.message('#%s' % userData[1], "You were given %d free %s packs by an admin. Check them using !freepacks ." % (amount, args[1]), True)
```
Wording/Grammar: Remove space before period at the end.

Dabomstew replied:
Yeah, removing the period would probably be the best solution overall for these. You nailed what I was thinking with the space.

xorhash commented on main.py:
```diff
@@ -1391,7 +1410,7 @@ def openBooster(userid, username, channel, isWhisper, packname, buying=True):
         # alerts
         for w in alertwaifus:
-            threading.Thread(target=sendDrawAlert, args=(channel, w, str(username))).start()
+            threading.Thread(target=sendDrawAlert, args=(channel, w, str(display_name))).start()
```
Sneaky display name change here. This may make it very difficult for people to check the puller's hand/profile to try and work out a deal immediately (if needed). I'd suggest changing sendDrawAlert to be like "{display_name} ({username})".

Dabomstew replied:
This was actually the display name all along. This function never used to pull in the username, but it now has to for the meme pack system. However, adjusting it like that (at least if the names differ by more than case) is probably a good idea regardless.

xorhash commented on main.py:
```diff
@@ -1680,6 +1706,23 @@ def timer():
+                    ymdNow = now.strftime("%Y-%m-%d")
+                    if ymdNow != config["last_pudding_check"]:
+                        config["last_pudding_check"] = ymdNow
+                        cur.execute("UPDATE config SET value = %s WHERE name = 'last_pudding_check'", [ymdNow])
```
Is there any particular reason you're doing a check based on textual dates, rather than just using the omnipresent `current_milli_time()` and just clamping the timestamp to be accurate to a day?

Dabomstew replied:
For once I want this code to be run once a day, not once every 24 hours. While I'm not sure if any relevant country can have its DST change over on the 1st or the 8th, it seems worth it to not risk anything.

xorhash replied:
Which reminds me: Can this double-process on leap seconds? I honestly don't know enough about leap seconds to tell, but I'm not sure what result you will get at 23:59:60.

xorhash commented on main.py:
```diff
@@ -1680,6 +1706,23 @@ def timer():
+                            cur.execute("UPDATE users SET puddingExpiring = puddingPrevious, puddingPrevious = puddingCurrent, puddingCurrent = 0")
+                            cur.execute("SELECT name, puddingExpiring FROM users WHERE puddingExpiring > 0")
+                            for userRow in cur.fetchall():
+                                self.message("#%s" % userRow[0], "You have %d pudding expiring on the 8th of this month. !pudding to see your balance and to spend it." % userRow[1], True)
```
"8th of this month" – you avoid hardcoding the name of rarities, which are arguably much more fundamental, but then hardcode the expiry day of month?

xorhash commented on main.py:
```diff
@@ -2154,12 +2158,54 @@ def do_command(self, command, args, sender, channel, tags, isWhisper=False):
+                            self.message(channel, "%s, you open a %s booster for %d pudding: %s/booster?user=%s" % (tags['display-name'], booster[0], cost, config["siteHost"], sender), isWhisper)
+                        except InvalidBoosterException:
+                            self.message(channel,
+                                        "Go tell an admin that this booster type is broken.",
```
Is there a reason to make the user deal with this? You now have a way to notify admins via Discord.

Telling an admin may an issue for casual players who may not know any admin's name.

Dabomstew replied:
This is just a generic message (since this error should almost never be thrown barring a huge misconfiguration) but that is a worthwhile point.

xorhash commented on main.py:
```diff
@@ -3695,10 +3695,58 @@ def do_command(self, command, args, sender, channel, tags, isWhisper=False):
                     config["promoschanged"] = "yes"
                     cur.execute("REPLACE INTO config(name, value) VALUES('promoschanged', 'yes')")
                 return
+            if command == "freepacks" or (command == "bet" and len(args) > 0 and args[0].lower() == "packs"):
```
This just occurred to me. You may want to alias this to `freepack` as well to account for people thinking "I'll open *a* free pack", similar how you have a singular version for `!sets` with `!set`.

Dabomstew commented on main.py:
```diff
@@ -2318,14 +2315,19 @@ def do_command(self, command, args, sender, channel, tags, isWhisper=False):
+                                    messageForHandUpgrade(tags['user-id'], tags['display-name'], self, channel, isWhisper)
+                                self.message(channel, "%s, you got your daily free reward: %sa %s booster - %s/booster?user=%s" % (tags['display-name'], pointsPrefix, rewardInfo[3], config['siteHost'], sender), isWhisper)
+                            except InvalidBoosterException:
+                                self.message(channel, "Oops! The free reward database appears to be misconfigured. Please report this to an admin.", isWhisper)
```
Yeah, different wording. Will clean these up in a commit later

xorhash commented on main.py:
```diff
@@ -5031,6 +5026,102 @@ def do_command(self, command, args, sender, channel, tags, isWhisper=False):
+                                cur.execute("UPDATE has_waifu SET custom_image = %s WHERE waifuid = %s AND userid = %s AND rarity = %s", [hostedURL, tags['user-id'], waifuid, godRarity])
+
+                            # log the change for posterity
+                            insertArgs = [tags['user-id'], waifuid, args[2], do_global, tags['user-id'], current_milli_time()]
```
You're using `tags['user-id']` for both the `requesterid` and `moderatorid` fields? Is this really what you meant?
